core/src/main/java/org/mule/processor/NonBlockingProcessorExecutor.java

```diff
@@ -0,0 +1,124 @@
+import org.slf4j.LoggerFactory;
+
+/**
+ * Specialized {@link org.mule.processor.BlockingProcessorExecutor} that pauses iteration in the case a {@link org
+ * .mule.processor.NonBlockingMessageProcessor}
+ * is invoked and the flow is executing using non-blocking.  Processor execution is then continued when the {@link
+ * org.mule.processor.NonBlockingMessageProcessor}
+ * invokes the {@link org.mule.api.transport.ReplyToHandler}.
+ */
+public class NonBlockingProcessorExecutor extends BlockingProcessorExecutor
+{
+
+    private static final Logger logger = LoggerFactory.getLogger(NonBlockingProcessorExecutor.class);
+    private final ReplyToHandler replyToHandler;
+    private volatile int index;
+
+    public NonBlockingProcessorExecutor(MuleEvent event, List<MessageProcessor> processors,
+                                        MessageProcessorExecutionTemplate executionTemplate, boolean copyOnVoidEvent)
+    {
+        super(event, processors, executionTemplate, copyOnVoidEvent);
+        this.replyToHandler = event.getReplyToHandler();
+    }
+
+    @Override
+    protected void preProcess(MessageProcessor processor)
+    {
+        if (event.isAllowNonBlocking() && replyToHandler != null)
+        {
+            if ((processor instanceof MessageRouter || processor instanceof InterceptingMessageProcessor) && !
+                    (processor instanceof NonBlockingSupported))
+            {
+                if (logger.isInfoEnabled())
+                {
+                    logger.info("The message processor {} does not currenlty support non-blocking execution and processing " +
+                                "will now fall back to blocking.  The 'non-blocking' processing strategy is not " +
+                                "recommened if unsupported message processors are being used.  ", processor);
+                }
+                event = new DefaultMuleEvent(event.getMessage(), event, true);
+            }
+
+            if (processor instanceof NonBlockingMessageProcessor)
+            {
+                event = new DefaultMuleEvent(event, new CompletionHandlerReplyToHandlerAdaptor(new NonBlockingProcessingCompletionHandler()));
+            }
+        }
+    }
+
+    private void resume() throws MessagingException
+    {
+        MuleEvent result = execute();
+        if (result != null && !(result instanceof VoidMuleEvent))
+        {
+            try
+            {
+                replyToHandler.processReplyTo(result, null, null);
+            }
+            catch (MuleException e)
+            {
+                replyToHandler.processExceptionReplyTo(event, new MessagingException(event, e), null);
+            }
+        }
+    }
+
+    @Override
+    protected int getIndex()
+    {
+        return index;
+    }
+
+    @Override
+    protected int incrementIndex()
+    {
+        return index++;
+    }
+
+    class NonBlockingProcessingCompletionHandler implements CompletionHandler<MuleEvent, MessagingException>
```

Comment:
I would rather make this private, this is internal to NonBlockingProcessorExecutor which is already tested.